Added SSE Parameter - 0.5

jrs526 opened the pull request:
Enables server-side encryption by adding a parameter, sse, which takes a string denoting the encryption algorithm.

jrs526 commented:
I haven't tested this yet, but wanted to put this out there so folks knew it existed.

donny-dont commented on main.go:
```diff
@@ -77,6 +77,11 @@ func main() {
+		cli.StringFlag{
+			Name:   "sse",
+			Usage:  "sse algorithm",
+			EnvVar: "PLUGIN_SSE",
```
Should there be a default here?

tboerger replied:
I think empty is the default :D

donny-dont replied:
Right but you need to do like this https://github.com/drone-plugins/drone-s3/blob/master/plugin.go#L88 rather than setting the value.

jrs526 replied:
Unless I'm missing something, I don't think we need to do anything like that here. This change only addresses the simplest use case, which only requires a string value for the SSECustomerAlgorithm.

donny-dont commented:
If you can check with Minio and S3 that'd work for me.

tboerger commented:
> If you can check with Minio and S3 that'd work for me.

Does minio already support that?

bradrydzewski commented on .gitignore:
```diff
@@ -26,3 +26,5 @@ _testmain.go
 coverage.out
 drone-s3
+
+*.swp
```
maybe should be removed and added to your `.git/info/exclude` since it seems machine specific?

jrs526 replied:
Sure.

donny-dont commented:
Is `sse` really a good name? Maybe `encrypt` would be better? `sse` is just vague.

jrs526 commented:
> Is sse really a good name? Maybe encrypt would be better? sse is just vague.

The `aws cp` command supports a `--sse` flag that takes a string value representing the algorithm to be used.  So, calling the parameter `sse` was an attempt at being consistent with the cli so that folks could easily translate their working aws commands to the plugin.

donny-dont commented:
I get that its just something at first glance I didn't know. Seeing `sse` makes me think SIMD stuff but that's possibly just me. Just throwing it out there.

jrs526 commented:
> I get that its just something at first glance I didn't know. Seeing sse makes me think SIMD stuff but that's possibly just me. Just throwing it out there.

Good point.  I'm sure you are not the only one that would make that connection.  I could certainly go with `encrypt` and update the docs to clarify the relationship to the `--sse` flag.

jrs526 commented:
@donny-dont I've renamed the parameter to `encryption` and squashed my commits.  I also noticed I was using the wrong field for the PutObject operation, so I fixed that.  Thoughts?

bradrydzewski commented on plugin.go:
```diff
@@ -153,11 +157,12 @@ func (p *Plugin) Exec() error {
+			Key:                  &target,
+			ACL:                  &(p.Access),
+			ContentType:          &content,
+			ServerSideEncryption: &(p.Encryption),
```
the aws client checks to see if `ServerSideEncryption` is not null. This will cause the aws client to think it is always enabled, since it is creating a pointer to an empty string by default which will evaluate to non-null

basically looks like this in the aws library:

```
if objData.ServerSideEncryption != nil {
	obj.Encrypted = true
	obj.EncryptionType = *objData.ServerSideEncryption
}
```

therefore I recommend adjusting this code to pass nil by default:

```
obj := &s3.PutObjectInput{
  Body:        f,
  Bucket:      &(p.Bucket),
  Key:         &target,	
  ACL:         &(p.Access),
  ContentType: &content,
}

if p.Encryption != "" {
  obj.ServerSideEncryption = &(p.Encryption)
}

_, err = client.PutObject(obj)
```

jrs526 commented:
Made @bradrydzewski's suggested changes.  Ran test uploads without encryption, with AES256, and with aws:kms.  All three worked.

bradrydzewski commented:
thanks!
